File: OpenPose/OpenPoseVideo.cpp
#include "pose_estimation.hpp"

#include <opencv2/highgui.hpp>
#include <opencv2/videoio.hpp>

#include <cmath>
#include <filesystem>
#include <iostream>
#include <stdexcept>
#include <string>

#ifndef OPENPOSE_SOURCE_DIR
#define OPENPOSE_SOURCE_DIR "."
#endif

namespace {

struct Options {
    std::filesystem::path input =
        std::filesystem::path(OPENPOSE_SOURCE_DIR) / "sample_video.mp4";
    std::filesystem::path model =
        std::filesystem::path(OPENPOSE_SOURCE_DIR) /
        "models" /
        "pose_estimation_mediapipe_2023mar.onnx";
    std::filesystem::path output_dir =
        std::filesystem::path(OPENPOSE_SOURCE_DIR) / "output";
    std::string device = "cpu";
    float score_threshold = 0.5F;
    int max_frames = 0;
    bool display = false;
    bool validate = false;
    bool help = false;
};

std::string requireValue(int& index, int argc, char** argv) {
    if (index + 1 >= argc) {
        throw std::invalid_argument(
            "Missing value after " + std::string(argv[index]));
    }
    ++index;
    return argv[index];
}

Options parseOptions(int argc, char** argv) {
    Options options;
    for (int index = 1; index < argc; ++index) {
        const std::string argument = argv[index];
        if (argument == "--help" || argument == "-h") {
            options.help = true;
        } else if (argument == "--input") {
            options.input = requireValue(index, argc, argv);
        } else if (argument == "--model") {
            options.model = requireValue(index, argc, argv);
        } else if (argument == "--output-dir") {
            options.output_dir = requireValue(index, argc, argv);
        } else if (argument == "--device") {
            options.device = requireValue(index, argc, argv);
        } else if (argument == "--score-threshold") {
            options.score_threshold =
                std::stof(requireValue(index, argc, argv));
        } else if (argument == "--max-frames") {
            options.max_frames =
                std::stoi(requireValue(index, argc, argv));
        } else if (argument == "--display") {
            options.display = true;
        } else if (argument == "--no-display") {
            options.display = false;
        } else if (argument == "--validate") {
            options.validate = true;
        } else {
            throw std::invalid_argument("Unknown argument: " + argument);
        }
    }
    if (options.score_threshold < 0.0F || options.score_threshold > 1.0F) {
        throw std::invalid_argument(
            "--score-threshold must be between 0 and 1.");
    }
    if (options.max_frames < 0) {
        throw std::invalid_argument("--max-frames cannot be negative.");
    }
    return options;
}

void printHelp(const char* executable) {
    std::cout
        << "Usage: " << executable << " [options]\n"
        << "  --input PATH             Input video\n"
        << "  --model PATH             MediaPipe Pose ONNX model\n"
        << "  --output-dir PATH        Directory for pose-video.avi\n"
        << "  --device cpu|cuda        DNN execution device\n"
        << "  --score-threshold VALUE  Landmark probability threshold\n"
        << "  --max-frames N           Zero processes the complete video\n"
        << "  --display                Open an interactive window\n"
        << "  --no-display             Run headlessly (default)\n"
        << "  --validate               Check stable output invariants\n";
}

void validateWrittenVideo(
    const std::filesystem::path& output_path,
    const cv::Size& expected_size,
    int expected_frames) {
    cv::VideoCapture capture(output_path.string());
    if (!capture.isOpened()) {
        throw std::runtime_error(
            "Could not reopen output video: " + output_path.string());
    }
    const cv::Size actual_size{
        cvRound(capture.get(cv::CAP_PROP_FRAME_WIDTH)),
        cvRound(capture.get(cv::CAP_PROP_FRAME_HEIGHT)),
    };
    const int actual_frames =
        cvRound(capture.get(cv::CAP_PROP_FRAME_COUNT));
    capture.release();
    if (actual_size != expected_size) {
        throw std::runtime_error("Output video dimensions changed.");
    }
    if (actual_frames != expected_frames) {
        throw std::runtime_error(
            "Output video frame count does not match processed frames.");
    }
}

}  // namespace

int main(int argc, char** argv) {
    try {
        const Options options = parseOptions(argc, argv);
        if (options.help) {
            printHelp(argv[0]);
            return 0;
        }

        cv::VideoCapture capture(options.input.string());
        if (!capture.isOpened()) {
            throw std::runtime_error(
                "Could not open input video: " + options.input.string());
        }

        const cv::Size frame_size{
            cvRound(capture.get(cv::CAP_PROP_FRAME_WIDTH)),
            cvRound(capture.get(cv::CAP_PROP_FRAME_HEIGHT)),
        };
        if (frame_size.width <= 0 || frame_size.height <= 0) {
            throw std::runtime_error("Input video has invalid dimensions.");
        }
        double fps = capture.get(cv::CAP_PROP_FPS);
        if (!std::isfinite(fps) || fps <= 0.0) {
            fps = 25.0;
        }

        std::filesystem::create_directories(options.output_dir);
        const std::filesystem::path output_path =
            options.output_dir / "pose-video.avi";
        cv::VideoWriter writer(
            output_path.string(),
            cv::VideoWriter::fourcc('M', 'J', 'P', 'G'),
            fps,
            frame_size);
        if (!writer.isOpened()) {
            throw std::runtime_error(
                "Could not create output video: " + output_path.string());
        }

        learnopencv::pose::PoseEstimator estimator(
            options.model, options.device);
        int processed = 0;
        long long total_visible = 0;
        long long total_edges = 0;
        cv::Mat frame;
        while (capture.read(frame)) {
            if (frame.empty()) {
                throw std::runtime_error(
                    "Decoded an empty frame at index " +
                    std::to_string(processed) + ".");
            }
            if (frame.size() != frame_size) {
                throw std::runtime_error(
                    "A decoded frame changed dimensions.");
            }

            const learnopencv::pose::PoseResult result =
                estimator.infer(frame);
            cv::Mat output = frame.clone();
            const learnopencv::pose::DrawMetrics metrics =
                learnopencv::pose::draw(
                    output, result, options.score_threshold);
            if (options.validate) {
                learnopencv::pose::validate(frame, result, metrics);
            }
            writer.write(output);
            ++processed;
            total_visible += metrics.visible_count;
            total_edges += metrics.edge_count;

            if (options.display) {
                cv::imshow("MediaPipe Pose", output);
                if (cv::waitKey(1) == 27) {
                    break;
                }
            }
            if (options.max_frames > 0 &&
                processed >= options.max_frames) {
                break;
            }
        }

        capture.release();
        writer.release();
        if (options.display) {
            cv::destroyAllWindows();
        }
        if (processed == 0) {
            throw std::runtime_error("No frames were decoded.");
        }

        if (options.validate) {
            validateWrittenVideo(output_path, frame_size, processed);
            std::cout
                << "VALIDATION PASSED: frames=" << processed
                << " size=" << frame_size.width
                << 'x' << frame_size.height << '\n';
        }

        std::cout << "OpenCV version: " << CV_VERSION << '\n';
        std::cout
            << "POSE VIDEO RESULT: frames=" << processed
            << " total_visible=" << total_visible
            << " total_edges=" << total_edges << '\n';
        std::cout << "Saved output: " << output_path << '\n';
        return 0;
    } catch (const std::exception& error) {
        std::cerr << "ERROR: " << error.what() << '\n';
        return 2;
    }
}

File: OpenPose/OpenPoseVideo.py
#!/usr/bin/env python3
"""Estimate one person's MediaPipe Pose landmarks throughout a video."""

from __future__ import annotations

import argparse
from pathlib import Path
import sys

import cv2

from pose_estimation import (
    DEFAULT_MODEL,
    PROJECT_DIR,
    draw_pose,
    infer_pose,
    load_pose_model,
    validate_pose,
)


DEFAULT_INPUT = PROJECT_DIR / "sample_video.mp4"


def parse_args(argv: list[str] | None = None) -> argparse.Namespace:
    """Expose deterministic video and headless test controls."""

    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--input",
        "--video_file",
        dest="input",
        type=Path,
        default=DEFAULT_INPUT,
        help=f"Input video (default: {DEFAULT_INPUT})",
    )
    parser.add_argument("--model", type=Path, default=DEFAULT_MODEL)
    parser.add_argument(
        "--output-dir",
        type=Path,
        default=PROJECT_DIR / "output",
        help="Directory for pose-video.avi",
    )
    parser.add_argument("--device", choices=("cpu", "cuda"), default="cpu")
    parser.add_argument("--score-threshold", type=float, default=0.5)
    parser.add_argument(
        "--max-frames",
        type=int,
        default=0,
        help="Stop after this many frames; zero processes the complete video.",
    )
    display_group = parser.add_mutually_exclusive_group()
    display_group.add_argument("--display", action="store_true")
    display_group.add_argument(
        "--no-display",
        action="store_true",
        help="Run headlessly (the default; accepted explicitly for CI).",
    )
    parser.add_argument("--validate", action="store_true")
    return parser.parse_args(argv)


def _validate_written_video(
    output_path: Path,
    expected_size: tuple[int, int],
    expected_frames: int,
) -> None:
    """Reopen the completed video and verify its dimensions and frame count."""

    capture = cv2.VideoCapture(str(output_path))
    if not capture.isOpened():
        raise RuntimeError(f"Could not reopen output video: {output_path}")
    width = int(round(capture.get(cv2.CAP_PROP_FRAME_WIDTH)))
    height = int(round(capture.get(cv2.CAP_PROP_FRAME_HEIGHT)))
    frame_count = int(round(capture.get(cv2.CAP_PROP_FRAME_COUNT)))
    capture.release()
    if (width, height) != expected_size:
        raise RuntimeError(
            f"Output size {(width, height)} does not match {expected_size}."
        )
    if frame_count != expected_frames:
        raise RuntimeError(
            f"Output frame count {frame_count} does not match {expected_frames}."
        )


def run(args: argparse.Namespace) -> dict[str, object]:
    """Process a video without discarding the first frame or reading past EOF."""

    input_path = args.input.expanduser().resolve()
    if args.max_frames < 0:
        raise ValueError("--max-frames cannot be negative.")
    if not 0.0 <= args.score_threshold <= 1.0:
        raise ValueError("--score-threshold must be between 0 and 1.")

    capture = cv2.VideoCapture(str(input_path))
    if not capture.isOpened():
        raise FileNotFoundError(f"Could not open input video: {input_path}")

    width = int(round(capture.get(cv2.CAP_PROP_FRAME_WIDTH)))
    height = int(round(capture.get(cv2.CAP_PROP_FRAME_HEIGHT)))
    fps = float(capture.get(cv2.CAP_PROP_FPS))
    if width <= 0 or height <= 0:
        capture.release()
        raise RuntimeError(f"Input video has invalid dimensions: {input_path}")
    if not fps or fps <= 0.0:
        fps = 25.0

    output_dir = args.output_dir.expanduser().resolve()
    output_dir.mkdir(parents=True, exist_ok=True)
    output_path = output_dir / "pose-video.avi"
    writer = cv2.VideoWriter(
        str(output_path),
        cv2.VideoWriter_fourcc(*"MJPG"),
        fps,
        (width, height),
    )
    if not writer.isOpened():
        capture.release()
        raise RuntimeError(f"Could not create output video: {output_path}")

    net = load_pose_model(args.model, args.device)
    processed = 0
    total_visible = 0
    total_edges = 0
    try:
        while True:
            has_frame, frame = capture.read()
            if not has_frame:
                break
            if frame is None or frame.size == 0:
                raise RuntimeError(f"Decoded an empty frame at index {processed}.")
            if frame.shape[1] != width or frame.shape[0] != height:
                raise RuntimeError(
                    f"Frame {processed} changed size from {(width, height)} "
                    f"to {(frame.shape[1], frame.shape[0])}."
                )

            result = infer_pose(net, frame)
            output, visible_count, edge_count = draw_pose(
                frame, result, args.score_threshold
            )
            if args.validate:
                validate_pose(frame, result, visible_count, edge_count)
            writer.write(output)
            processed += 1
            total_visible += visible_count
            total_edges += edge_count

            if args.display and not args.no_display:
                cv2.imshow("MediaPipe Pose", output)
                if cv2.waitKey(1) == 27:
                    break
            if args.max_frames and processed >= args.max_frames:
                break
    finally:
        capture.release()
        writer.release()
        if args.display:
            cv2.destroyAllWindows()

    if processed == 0:
        raise RuntimeError(f"No frames were decoded from input video: {input_path}")
    if args.validate:
        _validate_written_video(output_path, (width, height), processed)
        print(
            "VALIDATION PASSED: "
            f"frames={processed} size={width}x{height}"
        )

    print(f"OpenCV version: {cv2.__version__}")
    print(
        "POSE VIDEO RESULT: "
        f"frames={processed} total_visible={total_visible} "
        f"total_edges={total_edges}"
    )
    print(f"Saved output: {output_path}")
    return {
        "output": output_path,
        "frames": processed,
        "size": (width, height),
        "total_visible": total_visible,
        "total_edges": total_edges,
    }


def main(argv: list[str] | None = None) -> int:
    """Return a clear nonzero status for missing inputs, models, or writers."""

    try:
        run(parse_args(argv))
    except (FileNotFoundError, RuntimeError, ValueError, cv2.error) as error:
        print(f"ERROR: {error}", file=sys.stderr)
        return 2
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
